packages/material-ui/src/NotchedOutline/NotchedOutline.js

```diff
@@ -0,0 +1,207 @@
+import React from 'react';
```

Comment:
@oliviertassinari thank you for your review!

Yes, this is something I had considered initially. I went away from this after seeing Google's implementation using an svg. I felt that the svg implementation provided a better developer experience by not enforcing a background color, which could be difficult or potentially impossible to determine in certain cases. For example, my use case includes a background with a gradient, which would not be possible with this implementation.

I was also partial to Matt's animation of the notch opening 😄

I am curious if there is something else we could do to see the benefits of both approaches. For example, the `Resize-Observer` [uses a `MutationObserver` under the hood](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver). Pending that, it falls back to a less performant mechanism, but [it looks like this is well supported and the fallback is not necessary](https://caniuse.com/#feat=mutationobserver). I think that rewriting using this abstraction instead would help to solve at least the increased bundle size.

Alternatively, the `NotchedOutline` could be provided as a component which is not integrated into the `TextField`, thus allowing consumers to use this component by composing the `FormControl` components, and to manage the size of this as needed. This would allow consumers to use a library such as `Resize-Observer` to resize the component. This, however, is not a great developer experience.

@mbrookes any thoughts? I know you had a similar implementation of this component

Reply:
@enagy27 I had also wondered about composability - it removes some of the component the overhead for those that don't use the outlined variant. I didn't look into it any further though, as I took a break from my implementation, and yours took the lead, but I can imagine so challenges.

I'm not keen on the floating label background color approach - that was my first implementation. It works, but I'd consider it a bit of a fudge. Also, CSS border-width can't be animated, in addition to losing the notch animation as you mentioned.

Reply:
@mbrookes @oliviertassinari 

My thought is to add an `Outline` or `Decorator` render prop to the TextField. From there I will create an `OutlinedTextField` component and see how that takes shape. I think I'll give this a shot if it sounds like a good API to you guys? Thanks!

Reply:
> felt that the svg implementation provided a better developer experience by not enforcing a background color, which could be difficult or potentially impossible to determine in certain cases.

Take the `fieldset` + `legend` DOM elements. It behaves like the SVG implementation. I agree that it's easier to use. But I'm happy to make sacrifices like this if the Outline usage share is 20% and the customized text field usage share is 50%. I have a very limited use case of the Material Spec at work. It's really about customization. I want the problem that we solve to scale to different UIs. If making sacrifizes to the Material Design specification is the answer, so be it. 
https://github.com/material-components/material-components-web-react will do a good job at making a different tradeoff, more biased toward Material Design and less toward customization.

Anyway, I'm only writting this down so you understand my bias.

Reply:
@oliviertassinari I really appreciate your explanation, thank you! That absolutely makes sense. Sounds like two sides of the same coin.

`fieldset` + `legend` is a great idea! I wasn't familiar with these components and hadn't thought there was another way. Is this something you'd prefer strongly enough that you'd like me to take another stab at it using these components?

Some quick challenges so far using this approach:
* Haven't yet found a property to animate on for the opening 🤔
* There is a gap between the `fieldset` and the top of the input. Unsure if this is an acceptable tradeoff or a workaround

Reply:
> fieldset + legend is a great idea! I

@enagy27 It was only to illustrate that the SVG implementation matches how the browsers handle `fieldset`. I'm fine with keeping the SVG implementation, it's just a matter of finding a good tradeoff with the API to better support this use case and to reduce the overhead for people not having it.